src/main/java/com/epam/javacourses2016/task18/SolverTask18.java

```diff
@@ -11,26 +13,103 @@
      * @return Подматрица, состоящая из максимального количества одинаковых элементов.
      */
     RectangularIntegerMatrix getMaxSubMatrix(RectangularIntegerMatrix matrix) {
-        //TODO
-        return null;
+        Stack<Matrix> matrixStack = getStackSubMatrix(matrix);
+        Matrix maxSubMatrix = null;
+        if (!matrixStack.empty()) {
+            maxSubMatrix = matrixStack.peek();
+            int maxNumberOfIdenticalValues = numberOfIdenticalValue(maxSubMatrix);
+            for (Matrix subMatrix : matrixStack) {
+                int numberOfIdenticalValues = numberOfIdenticalValue(subMatrix);
+                if (numberOfIdenticalValues > maxNumberOfIdenticalValues) {
+                    maxNumberOfIdenticalValues = numberOfIdenticalValues;
+                    maxSubMatrix = subMatrix;
+                }
+            }
+        }
+        return maxSubMatrix;
+    }
+
+    public Matrix getSubMatrix(RectangularIntegerMatrix matrix, int leftUpIndexRow, int leftUpIndexCol, int rightDownIndexRow, int rightDownIndexCol) throws NegativeArraySizeException {
+        int[][] subMatrix = new int[rightDownIndexRow - leftUpIndexRow][rightDownIndexCol - leftUpIndexCol];
+        for (int i = 0; i < subMatrix.length; i++) {
+            for (int j = 0; j < subMatrix[i].length; j++) {
+                subMatrix[i][j] = matrix.getValue(leftUpIndexRow + i, leftUpIndexCol + j);
+            }
+        }
+        return new Matrix(subMatrix);
+    }
+
+    public Stack<Matrix> getStackSubMatrix(RectangularIntegerMatrix matrix) {
+        Stack<Matrix> matrixStack = new Stack<>();
+        for (int x = 0; x < matrix.getHeight(); x++) {
+            for (int y = 0; y < matrix.getWidth(); y++) {
+                for (int i = 0; i < matrix.getHeight(); i++) {
+                    for (int j = 0; j < matrix.getWidth(); j++) {
```

Comment:
four for-sections looks terrified) If there is no way to exclude them, separate them in methods